### PR TITLE
Resolve #1432: Merge online indexer sub modules - "by records" and 'm…

### LINKED ISSUE #1432: Merge online indexer sub modules - "by records" and 'multi target by records"
After gaining some confidence in the "multi target by records" indexing path, change the code to let the "single target by records" be a private case of one item in the target index list.

### PR DESCRIPTION
**Pull request by jjezra**
…ulti target by records"
At this step, we still keep the old `IndexingByRecords` module - used for implementing `buildRange`, `splitIndexBuildRange`, `buildUnbuiltRange`,`buildEndpoints`. Maybe it'll be possible to delete it after implementing parallel indexing in the new modules.

### CODE REVIEW

**Comment by jjezra** [comment]
Test this please

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java [hunk 1]
```diff
@@ -192,7 +191,7 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
                                           lastResult.get().get().getIndexEntry().getKey() :
                                           rangeEnd)
                     .thenCompose(cont -> rangeSet.insertRange(store.ensureContextActive(), packOrNull(rangeStart), packOrNull(cont), true)
-                                .thenApply(ignore -> !Objects.equals(cont, rangeEnd)));
+                                .thenApply(ignore -> !allRangesExhausted(cont, rangeEnd)));
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java:194]: This feels like a somewhat weird change to make. It would seem to assume that `rangeEnd` is always `null`. If that's true, we should probably change the method to reject non-`null` values to `rangeEnd`.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java [hunk 2]
```diff
@@ -165,8 +198,7 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
                                           lastResult.get().get().getPrimaryKey() :
                                           rangeEnd)
                     .thenCompose(cont -> insertRanges(store.ensureContextActive(), targetRangeSets, packOrNull(rangeStart), packOrNull(cont))
-                                .thenApply(ignore -> !Objects.equals(cont, rangeEnd)));
-
+                                .thenApply(ignore -> !allRangesExhausted(cont, rangeEnd)));
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:201]: This also feels like maybe it works in practice because the wrong kind of range is never put into the range set...but that feels a little delicate to me. It's possible I'm missing part of some bigger picture, but it seems like we'd want to retain the ability in general to have a non-contiguous range built so that we can split up the records subspace and build the index in parallel, and a change like this would make that more difficult

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java [hunk 3]
```diff
@@ -202,8 +238,7 @@ CompletableFuture<Void> rebuildIndexInternalAsync(FDBRecordStore store) {
 
     @Nonnull
     @SuppressWarnings("PMD.CloseResource")
-    private CompletableFuture<Tuple> rebuildRangeOnly(@Nonnull FDBRecordStore store, Tuple cont, @Nonnull AtomicLong recordsScanned) {
-
+    private CompletableFuture<Tuple> rebuildRangeOnly(@Nonnull FDBRecordStore store, Tuple cont, @Nonnull AtomicLong recordsScanned, @Nullable Tuple rangeEnd) {
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:241]: This API doesn't seem right to me. I _think_ I understand what's going on, which is that it is assuming that index builds either require:

1. Scanning all records, in which case `rangeEnd` is `null` or
2. Scanning a set of range of records to cover all indexed types, in which case `rangeEnd` is the type key of the lexicographically last type

The problem I have with it is that the `rangeEnd` tuple here is a little too divorced from its origin for that to be clear to the user, especially because in most of the rest of the API, a bare `Tuple` is not assumed to be an inclusive right endpoint.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java [hunk 4]
```diff
@@ -823,6 +823,12 @@ protected CompletableFuture<Void> iterateAllRanges(List<Object> additionalLogMes
                 getRunner().getExecutor());
     }
 
+    protected static boolean allRangesExhausted(Tuple cont, Tuple end) {
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java:826]: In each case this function is used, it seems to be baking in an implicit assumption that `end` will always be `null`, as otherwise, the iteration of the loops won't ever stop, but I'm not sure that that's true in all cases

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java [hunk 5]
```diff
@@ -77,7 +76,8 @@ public class IndexingByRecords extends IndexingBase {
 
     IndexingByRecords(@Nonnull IndexingCommon common, @Nonnull OnlineIndexer.IndexingPolicy policy) {
         super(common, policy);
-        this.recordsRange = computeRecordsRange();
+        final TupleRange range = common.computeRecordsRange();
+        this.recordsRange = range == null ? TupleRange.ALL : range;
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java:80]: Rather than retaining both `IndexingByRecords` and `IndexingMultiTargetByRecords` until the range based APIs can be implemented on both (or removed), would it instead make sense to try and isolate the logic that truly is different between single- and multi-index situations, and then delete one of the classes? From what I can tell, the differences are:

1. When computing the records range, all record types from all indexes need to be included so as to cover all needed records
2. When indexing, all index maintainers need to be updated
3. When updating index states, each index needs to be updated

There are maybe a few other things, but this PR seems to also touch code related to picking ranges to scan, which I think should actually be common between the indexing methods

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java [hunk 6]
```diff
@@ -102,20 +110,45 @@ CompletableFuture<Void> buildIndexInternalAsync() {
...
+            rangeEnd = range.end;
+        }
+
+        final CompletableFuture<FDBRecordStore> maybePresetRangeFuture =
+                rangeStart == null ?
+                CompletableFuture.completedFuture(null) :
+                buildCommitRetryAsync((store, recordsScanned) -> {
+                    // Here: only records inside the defined records-range are relevant to the index. Hence, the completing range
+                    // can be preemptively marked as indexed.
+                    final RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(common.getIndex()));
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:137]: Is it correct that only `common.getIndex()`'s index range subspace gets updated? That doesn't seem right for non-idempotent indexes, but maybe those are excluded from multi-target index builds?

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java (hunk 6 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:137]: That's a bug. Will fix.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java (hunk 1 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java:194]:
```
            if (Boolean.FALSE.equals(hasNext)) {
                return AsyncUtil.READY_FALSE; // no more missing ranges - all done
            }
            final Range range = ranges.next();
            final Tuple rangeStart = RangeSet.isFirstKey(range.begin) ? null : Tuple.fromBytes(range.begin);
            final Tuple rangeEnd = RangeSet.isFinalKey(range.end) ? null : Tuple.fromBytes(range.end);
```
It is always null at the last unbuilt range (if the loop hadn't stop few lines above for no ranges). What am I missing?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java (hunk 2 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:201]: I was hoping to avoid the use of predefined ranges that are provided by the users in parallel indexing. At the current indexing state, the iteration should stop only when the index is fully built.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java (hunk 3 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:241]: Would a comment and renaming help?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java (hunk 5 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java:80]: Sorry, but I'm not sure that I understand this comment. Can you clarify?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java (hunk 6 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java (hunk 5 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java:80]: (after re-reading) 
Yes, I do not wish to retain both modules but to (eventually) delete `IndexingByRecords`, or just leave it with minimal functionality for API backward compatibility - if we decide to keep it. 
The multi-target module also has a slightly different flow, which is simpler and more compatible with the other new modules. It's probably a good idea to sharing code when it's possible and easy - like `computeRecordsRange` here..

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java (hunk 4 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java:826]: I believe that the new condition effectively always assumes it is building the entire range (which is I suppose the "implicit assumption" that I thought it was "baking in"). That's wrong in the cases where we only want to build a range, though it sounds like we're trying to get rid of that in favor of an alternative approach anyway. But if that's the plan, I'd rather we just remove and deprecate those endpoints, possibly waiting until we have a better story around parallel index builds.

Now that I'm looking closer, I think the _original_ condition was also wrong, though maybe in a different way (I think). I'm mainly looking at `buildRangeOnly`, which is supposed to take a range, and ensure that that whole range is built. Before this PR, that code:

1. Reads the range set to compute unbuilt ranges in the given range then
2. Looks in the results and builds the range over multiple transactions until
3. It builds all the way until it reaches `rangeEnd`

However, `rangeEnd` there is a range from the range set, meaning it effectively only builds the first (contiguous) range of data from the target range. The desired behavior, I think, is that it build all the way until it gets to the end of the target range. So, for `buildRangeOnly`, I think a better condition is something like "until the tuple given matches a tuple constructed from `rangeEndBytes` (or is `null` if `rangeEndBytes` is null)".

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/TupleRange.java [hunk 7]
```diff
@@ -213,6 +213,18 @@ public static TupleRange between(@Nullable Tuple low, @Nullable Tuple high) {
         return new TupleRange(low, high, lowEndpoint, highEndpoint);
     }
 
+    /**
+     * Create a <code>TupleRange</code> over all keys between the given {@link Tuple}s, including both endpoints.
+     * @param low the inclusive start of the range
+     * @param high the exclusive end of the range
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/TupleRange.java:219]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/TupleRange.java [hunk 8]
```diff
@@ -213,6 +213,18 @@ public static TupleRange between(@Nullable Tuple low, @Nullable Tuple high) {
...
+     * @return a <code>TupleRange</code> between <code>start</code> and <code>end</code>, includes.
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/TupleRange.java:220]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java (hunk 4 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java:826]: (assume:
'old path' = `IndexingByRecords` indexer - which was kept "as is" as much as possible. I'll be happy to change it as little as possible until reaching the stage of deleting unused code. 
'new path' = `IndexingMultiTargetByRecords`, 'IndexingByIndex`, `IndexingScrub*` indexers - a simplified path with no endpoint)

I believe that the current stop condition of the 'new path' was, indeed, a bug. My bug. It was harmless because these new indexers could not never have caused or 'inherit' a fragmented rangeSet. The name of the new stop condition - `allRangesExhausted` - should clarify that it return true if no more missing ranges exist, which is what these new indexers expect (I think). When the time comes, a different stop condition (`thisRangeIsExhausted`?) can be used to support a single range indexing. 
Although the 'old path' is hard to follow, I believe that its stop condition is fine because we do have some fragmented indexing tests. While developing, these tests had immediately failed when `IndexingMultiTargetByRecords` was defaulted to index single targets.
I'm not sure that I understood your last paragraph. IIUIC, `buildRangeOnly`'s return value is the stop condition of `iterateAllRanges`'s loop. For the current 'new path' indexers it returns true until `allRangesExhausted` - which means either continue indexing the same range or move on to the next unbuilt one. Am I missing something?

**Comment by jjezra** [comment]
Rebased, resolved conflicts, and squashed...

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java (hunk 2 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:205]: My problem with that is that this is not the current API of the indexer, and so making these changes seems premature unless you _also_ remove the `startBytes` and `endBytes` from the API (or require that `startBytes` and `endBytes` are both `null` at runtime). Which could certainly be done, and I'm fine with merging these changes after removing the endpoints we don't want to support any more.

I think if we want to understand why the fragmented indexing tests failed with the initial version of this PR but succeeded with the change to `allRangesExhausted`, I think what's going on is that when there is a range in the range set/an end range passed to this method, `allRangesExhausted` will always return `false`, so it will keep looping, until such time as the range set has no more ranges within it, at which point the earlier check (commented "no more missing ranges") fires. This wasn't a problem with the single index indexer because it uses the `hasMore` in a different way from the multi-target case.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java (hunk 1 above)
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java:194]: This condition is one of the places where using `allRangesExhausted` looks wrong to me (both now and before this PR). Note that `rangeEnd` was computed from reading the `rangeSet` above, so this lambda will now never return `false` if there's every any range in the range set. Before this PR, it was also wrong, but it would return `false` as soon as it completes the first range. But I think it's working, in practice, because `IndexingByIndex` builds would always be structured in a way that they'd have a `null` value as the end of at least one range. The way it reads the range set and only appears to grab the first element is....also kind of suspicious, though that's not affected by this change as directly

**Review by alecgrieser — CHANGES_REQUESTED** [review]

**Review by alecgrieser — APPROVED** [review]
